### PR TITLE
Unit test make regions

### PR DESCRIPTION
**Pull request by matheus220**
Turn RBR.makeRegions() package private and add unit tests to it.

Commit:Add

TaskNumber:#72

### CODE REVIEW

File: src/rbr/RBR.java [hunk 1]
```diff
@@ -149,7 +149,7 @@ void adjustRegions(Vertex sw) {
     }
 
     // Make regions only with correct destinations
```
Comment (jmferreiratech) [inline src/rbr/RBR.java]: Please, remove this commented line.

File: tests/rbr/MakeRegionsTest.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
+package rbr;
+
+import java.util.*;
+import org.junit.*;
+import util.*;
+
+public class MakeRegionsTest {
+
+    Set<Character> ip = new HashSet<>();
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: Couldn't these be private?

File: tests/rbr/MakeRegionsTest.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    public void emptyDestinationsCreatesNoRegion() throws Exception {
+        List<Region> teste = rbr.makeRegions(dst, ip, op);
+        Set<Vertex> emptyDst = new HashSet<>();
+        List<Region> createdRegionsWithoutDestinations = rbr.makeRegions(emptyDst, ip, op);
+        Assert.assertTrue(createdRegionsWithoutDestinations.isEmpty());
+    }
+
+    @Test
+    public void createdRegionsHaveSameInputPortsOriginals() throws Exception {
+        for(Region adjustedRegion : rbr.makeRegions(dst, ip, op)) {
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: 'adjustedRegion' doesn't seem a good name, does it?

File: tests/rbr/AdjustRegionsTest.java [hunk 4]
```diff
@@ -40,7 +40,7 @@ public void regionWithoutOutsidersIsntAdjusted() throws Exception {
         Assert.assertEquals(adjustedRegionWithoutOutsiders, arrayRegionWithoutOutsiders);
     }
 
-    @Ignore @Test
+    @Test
```
Comment (rafaelgmota) [inline tests/rbr/AdjustRegionsTest.java:43]: This modificaition is already done at pull request #95

File: tests/rbr/MakeRegionsTest.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Test
+    public void createdRegionsHaveSameOutputPortsOriginals() throws Exception {
+        for(Region adjustedRegion : rbr.makeRegions(dst, ip, op)) {
+            Assert.assertEquals(adjustedRegion.outputPorts(), op);
+        }
+    }
+
+    @Test
+    public void createdRegionsHaveNoOutsider() throws Exception {
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: Rename the tests that make assertions for each region to make it clear. This one I'd name 'createdRegionHasNoOutsider'.

File: tests/rbr/MakeRegionsTest.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    Set<Character> ip = new HashSet<>();
+    Set<Character> op = new HashSet<>();
+    Set<Vertex> dst = new HashSet<>();
+    private RBR rbr;
+
+    @Before
+    public void setUp() {
+        Graph graph = RegularGraphBuilder.generateGraph(4,4);
+        rbr = new RBR(graph);
+        ip.add('I');
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: Rename the variables to something less coded and that reflect multiplicity: 'ip' to 'inputPorts', 'op' to 'outputPorts', 'dst' to 'destinations' and so forth.

File: tests/rbr/MakeRegionsTest.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        dst.add(new Vertex("2.3"));
+        dst.add(new Vertex("3.0"));
+        dst.add(new Vertex("3.1"));
+        dst.add(new Vertex("3.2"));
+        dst.add(new Vertex("3.3"));
+    }
+
+    @Test
+    public void emptyDestinationsCreatesNoRegion() throws Exception {
+        List<Region> teste = rbr.makeRegions(dst, ip, op);
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: Do you need this 'teste' variable?

**Review by jmferreiratech — CHANGES_REQUESTED** [review]

File: tests/rbr/MakeRegionsTest.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    @Test
+    public void createdRegionsHaveSameDestinationsOriginals() throws Exception {
+        Set<Vertex> destinationsCreatedRegions = new HashSet<>();
+        Set<Vertex> originalsDst = new HashSet<>(dst);
+        for(Region createdRegion : rbr.makeRegions(originalsDst, ip, op)) {
+            destinationsCreatedRegions.addAll(createdRegion.destinations());
+        }
+        Assert.assertEquals(destinationsCreatedRegions, dst);
+    }
+}
```
Comment (rafaelgmota) [inline tests/rbr/MakeRegionsTest.java:86]: You can improve this by putting some test to verify if the number of created regions is correct and if the max and min of regions box is also correct.

**Review by rafaelgmota — CHANGES_REQUESTED** [review]

File: tests/rbr/AdjustRegionsTest.java (hunk 4 above)
Comment (jmferreiratech) [reply tests/rbr/AdjustRegionsTest.java:43]: Rebase your patch on top of upstream.

File: tests/rbr/MakeRegionsTest.java (hunk 8 above)
Comment (matheus220) [reply tests/rbr/MakeRegionsTest.java:86]: Is it not sufficient to verify that the set of destinations is the same as the original and that no region has outsiders?

Comment (jmferreiratech) [reply tests/rbr/MakeRegionsTest.java:86]: I think it's good to assert the number of generated regions.

**Review by jmferreiratech — CHANGES_REQUESTED** [review]
I think you can add two more useful tests: one that shows that each generated region box do not include the others generated regions' boxes. And another that shows that each generated region box isn't contiguous with any other generated region box.
These tests show that the method tries to minimize the number of generated regions.

**Comment by matheus220** [comment]
Ok, good idea. I'm working on it and I had a question. For a network with dimensions 4 x 4, with destinations placed alternately, similar to a chess. The minimum number of regions is 8?

**Comment by jmferreiratech** [comment]
Yes, if you have a 'chess pattern', the minimal number of regions is the number of destinations (as you do not have any vertex side-by-side).

File: tests/rbr/MakeRegionsTest.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Test
+    public void emptyDestinationsCreatesNoRegion() throws Exception {
+        List<Region> createdRegionsWithoutDestinations = rbr.makeRegions(new HashSet<>(), inputPorts, outputPorts);
+        Assert.assertTrue(createdRegionsWithoutDestinations.isEmpty());
+    }
+
+    @Test
+    public void createdRegionsHaveSameInputPortsOriginals() throws Exception {
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: Make the tests' name more 'sentence like'. Something as "createdRegionsHaveOriginalInputPorts".

File: tests/rbr/MakeRegionsTest.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        for(Region createdRegion : rbr.makeRegions(destinations, inputPorts, outputPorts)) {
+            Assert.assertEquals(createdRegion.inputPorts(), inputPorts);
+        }
+    }
+
+    @Test
+    public void createdRegionsHaveSameOutputPortsOriginals() throws Exception {
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: Idem

File: tests/rbr/MakeRegionsTest.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    public void createdRegionHasNoOutsider() throws Exception {
+        for(Region createdRegion : rbr.makeRegions(destinations, inputPorts, outputPorts)){
+            Assert.assertTrue(createdRegion.outsiders().isEmpty());
+        }
+    }
+
+    @Test
+    public void createdRegionsHaveSameDestinationsOriginals() throws Exception {
+        Set<Vertex> destinationsCreatedRegions = new HashSet<>();
+        Set<Vertex> originalsDst = new HashSet<>(destinations);
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: Do you really need to copy the destinations set?

File: tests/rbr/MakeRegionsTest.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Test
+    public void createdRegionHasNoOutsider() throws Exception {
+        for(Region createdRegion : rbr.makeRegions(destinations, inputPorts, outputPorts)){
+            Assert.assertTrue(createdRegion.outsiders().isEmpty());
+        }
+    }
+
+    @Test
+    public void createdRegionsHaveSameDestinationsOriginals() throws Exception {
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: I'd name it "createdRegionsCoverAllOriginalDestinations".

File: tests/rbr/MakeRegionsTest.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        Set<Vertex> destinationsCreatedRegions = new HashSet<>();
+        Set<Vertex> originalsDst = new HashSet<>(destinations);
+        List<Region> createdRegions = rbr.makeRegions(destinations, inputPorts, outputPorts);
+        for(Region createdRegion : createdRegions) {
+            destinationsCreatedRegions.addAll(createdRegion.destinations());
+        }
+        Assert.assertEquals(destinationsCreatedRegions, originalsDst);
+        Assert.assertEquals(createdRegions.size(), 5);
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: It does not make sense to assert that here. You can remove it by now. And then you'll not need the 'createdRegions' variable.

File: tests/rbr/MakeRegionsTest.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Test
+    public void createdRegionsAreDisjoint() throws Exception {
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: They do not need to be disjoint. Then, you can drop it.

File: tests/rbr/MakeRegionsTest.java [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        Set<Vertex> originalsDst = new HashSet<>(destinations);
+        for(Region createdRegion : rbr.makeRegions(destinations, inputPorts, outputPorts)) {
+            Assert.assertTrue(originalsDst.containsAll(createdRegion.destinations()));
+            originalsDst.removeAll(createdRegion.destinations());
+        }
+    }
+
+    @Test
+    public void createdMinimalNumberRegions() throws Exception {
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: Make the test name more 'sentence like'.

**Review by jmferreiratech — CHANGES_REQUESTED** [review]

File: tests/rbr/MakeRegionsTest.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+    public void createdRegionHasNoOutsider() throws Exception {
+        for(Region createdRegion : rbr.makeRegions(destinations, inputPorts, outputPorts)){
+            Assert.assertTrue(createdRegion.outsiders().isEmpty());
+        }
+    }
+
+    @Test
+    public void createdRegionsCoverAllOriginalDestinations() throws Exception {
+        Set<Vertex> destinationsCreatedRegions = new HashSet<>();
+        Set<Vertex> originalsDst = new HashSet<>(destinations);
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java]: You don't need 'originalsDst', do you? I think you can use 'destinations' directly.

File: tests/rbr/MakeRegionsTest.java [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    @Test
+    public void createdRegionHasNoOutsider() throws Exception {
+        for(Region createdRegion : rbr.makeRegions(destinations, inputPorts, outputPorts)){
+            Assert.assertTrue(createdRegion.outsiders().isEmpty());
+        }
+    }
+
+    @Test
+    public void createdRegionsCoverAllOriginalDestinations() throws Exception {
+        Set<Vertex> destinationsCreatedRegions = new HashSet<>();
```
Comment (jmferreiratech) [inline tests/rbr/MakeRegionsTest.java:67]: Instantiate the variable right before you use it.

**Review by jmferreiratech — CHANGES_REQUESTED** [review]
Make this changes and I'll merge it. Thanks.

**Comment by matheus220** [comment]
And this way?

**Review by jmferreiratech — APPROVED** [review]

**Pull request merged**